session/nontransactional.go

```diff
@@ -0,0 +1,509 @@
+	driver "github.com/pingcap/tidb/types/parser_driver"
+	"github.com/pingcap/tidb/util/collate"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"go.uber.org/zap"
+)
+
+// job: handle keys in [start, end]
+type job struct {
+	start   types.Datum
+	end     types.Datum
+	err     error
+	jobID   int
+	jobSize int
+	sql     string
+}
+
+// statementBuildInfo contains information that is needed to build the split statement in a job
+type statementBuildInfo struct {
+	stmt              *ast.NonTransactionalDeleteStmt
+	shardColumnType   types.FieldType
+	shardColumnRefer  *ast.ResultField
+	originalCondition ast.ExprNode
+}
+
+func (j job) String() string {
+	return fmt.Sprintf("job id: %d, job size: %d, range: [%s, %s]", j.jobID, j.jobSize, j.start.String(), j.end.String())
+}
+
+// HandleNonTransactionalDelete is the entry point for a non-transactional delete
+func HandleNonTransactionalDelete(ctx context.Context, stmt *ast.NonTransactionalDeleteStmt, se Session) (sqlexec.RecordSet, error) {
+	err := core.Preprocess(se, stmt)
+	if err != nil {
+		return nil, err
+	}
+	if !(se.GetSessionVars().IsAutocommit() && !se.GetSessionVars().InTxn()) {
+		return nil, errors.Errorf("non-transactional statement can only run in auto-commit mode. auto=commit:%v, inTxn:%v",
+			se.GetSessionVars().IsAutocommit(), se.GetSessionVars().InTxn())
+	}
+	tableName, selectSQL, shardColumn, err := buildSelectSQL(stmt, se)
+	if err != nil {
+		return nil, err
+	}
+	if stmt.DryRun == ast.DryRunQuery {
+		return buildDryRunResults(stmt.DryRun, []string{selectSQL}, se.GetSessionVars().BatchSize.MaxChunkSize)
+	}
+	jobs, err := buildShardJobs(ctx, stmt, se, selectSQL, shardColumn)
+	if err != nil {
+		return nil, err
+	}
+
+	splitStmts, err := splitDeleteWorker(ctx, jobs, stmt, tableName, se, stmt.DeleteStmt.Where)
+	if err != nil {
+		return nil, err
+	}
+	if stmt.DryRun == ast.DryRunSplitDml {
+		return buildDryRunResults(stmt.DryRun, splitStmts, se.GetSessionVars().BatchSize.MaxChunkSize)
+	}
+	return buildExecuteResults(jobs, se.GetSessionVars().BatchSize.MaxChunkSize)
+}
+
+// single-threaded worker. work on the key range [start, end]
+func splitDeleteWorker(ctx context.Context, jobs []job, stmt *ast.NonTransactionalDeleteStmt,
+	tableName *ast.TableName, se Session, originalCondition ast.ExprNode) ([]string, error) {
+
+	// prepare for the construction of statement
+	var shardColumnRefer *ast.ResultField
+	var shardColumnType types.FieldType
+	for _, col := range tableName.TableInfo.Columns {
+		if col.Name.L == stmt.ShardColumn.Name.L {
+			shardColumnRefer = &ast.ResultField{
+				Column: col,
+				Table:  tableName.TableInfo,
+			}
+			shardColumnType = col.FieldType
+		}
+	}
+	if shardColumnRefer == nil && stmt.ShardColumn.Name.O != "_tidb_rowid" {
+		return nil, errors.New("Non-transactional delete, column not found")
+	}
+
+	splitStmts := make([]string, 0, len(jobs))
+	for i := range jobs {
+		select {
+		case <-ctx.Done():
```

Comment:
Need to test if the `kill tidb xxx` statement  could stop the job processing as expected.

Reply:
I tested it manually. It works.